fish-rust/src/fish.rs

```diff
@@ -0,0 +1,877 @@
+        "     total time: {} ms",
+        tv_to_msec(&rs.ru_utime) + tv_to_msec(&rs.ru_stime)
+    );
+    let _ = writeln!(dest, "        max rss: {} kb", rss_kb);
+    let _ = writeln!(dest, "        signals: {}", rs.ru_nsignals);
+}
+
+fn determine_config_directory_paths(argv0: &wstr) -> ConfigPaths {
+    // PORTING: why is this not just an associated method on ConfigPaths?
+
+    let mut paths = ConfigPaths::default();
+    let mut done = false;
+    // FIXME: osstr
+    let exec_path = get_executable_path(&argv0.to_string());
+    if let Ok(exec_path) = exec_path.canonicalize() {
+        FLOGF!(
+            config,
+            format!("exec_path: {:?}, argv[0]: {:?}", exec_path, argv0)
+        );
+        // TODO: we should determine program_name from argv0 somewhere in this file
+
+        // Detect if we're running right out of the CMAKE build directory
+        if exec_path.starts_with(OUT_DIR) {
+            let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+            FLOGF!(
+                config,
+                "Running out of target directory, using paths relative to CARGO_MANIFEST_DIR:\n",
+                manifest_dir.display()
+            );
+            done = true;
+            paths = ConfigPaths {
+                data: manifest_dir.join("share"),
+                sysconf: manifest_dir.join("etc"),
+                doc: manifest_dir.join("user_doc/html"),
+                bin: OUT_DIR.into(),
+            }
+        }
+
+        if !done {
+            // The next check is that we are in a reloctable directory tree
+            let installed_suffix = Path::new("/bin/fish");
+            let just_a_fish = Path::new("/fish");
+            let suffix = if exec_path.ends_with(installed_suffix) {
+                Some(installed_suffix)
+            } else if exec_path.ends_with(just_a_fish) {
+                FLOG!(
+                    config,
+                    "'fish' not in a 'bin/', trying paths relative to source tree"
+                );
+                Some(just_a_fish)
+            } else {
+                None
+            };
+
+            if let Some(suffix) = suffix {
+                let seems_installed = suffix == installed_suffix;
+
+                let base_path: PathBuf = {
+                    // overly elaborate way to strip the suffix from base_path
+                    let mut remaining_suffix = suffix;
+                    let mut base = exec_path;
+                    while let Some(suffix) = remaining_suffix.parent() {
+                        base.pop();
+                        remaining_suffix = suffix;
+                    }
+                    // remove whatever remains in remaining_suffix
+                    base.pop();
+                    base
+                };
+
+                paths = if seems_installed {
+                    ConfigPaths {
+                        data: base_path.join("share/fish"),
+                        sysconf: base_path.join("etc"),
```

Comment:
this should be `/etc/fish` (if it's installed)

Reply:
Good catch! Will fix